File: hard-way/dp/322.零钱兑换.go
/*
 * @lc app=leetcode.cn id=322 lang=golang
 *
 * [322] 零钱兑换
 *
 * https://leetcode-cn.com/problems/coin-change/description/
 *
 * algorithms
 * Medium (34.10%)
 * Likes:    247
 * Dislikes: 0
 * Total Accepted:    24.8K
 * Total Submissions: 69.2K
 * Testcase Example:  '[1,2,5]\n11'
 *
 * 给定不同面额的硬币 coins 和一个总金额
 * amount。编写一个函数来计算可以凑成总金额所需的最少的硬币个数。如果没有任何一种硬币组合能组成总金额，返回 -1。
 *
 * 示例 1:
 *
 * 输入: coins = [1, 2, 5], amount = 11
 * 输出: 3
 * 解释: 11 = 5 + 5 + 1
 *
 * 示例 2:
 *
 * 输入: coins = [2], amount = 3
 * 输出: -1
 *
 * 说明:
 * 你可以认为每种硬币的数量是无限的。
 *
 */

// @lc code=start
import "math"

func min(a, b int) int {
	if a <= b {
		return a
	}
	return b
}

func coinChange(coins []int, amount int) int {
	// 拆分子问题时，会发现组合的多少之和当前币种的选择有关
	// f(n)=min{f(n-1),f(n-5),f(n-11)}+1

	dp := make([]int, amount+1)
	for i := 0; i < amount+1; i++ {
		dp[i] = math.MaxInt32
	}
	dp[0] = 0

	for i := 1; i <= amount; i++ {
		cost := math.MaxInt32
		for j := 0; j < len(coins); j++ {
			if i-coins[j] >= 0 {
				if dp[i-coins[j]] != math.MaxInt32 {
					cost = min(cost, dp[i-coins[j]]+1)
				}
			}
		}
		dp[i] = cost
	}
	if dp[amount] == math.MaxInt32 {
		return -1
	}
	return dp[amount]
}

// @lc code=end

File: labuladong/开始之前/dp/322.零钱兑换.py
#
# @lc app=leetcode.cn id=322 lang=python3
#
# [322] 零钱兑换
#

# @lc code=start
class Solution:
    def coinChange_with_memory(self, coins, amount: int) -> int:
        """这个题目是符合最优子结构的
        比如你想求 amount = 11 时的最少硬币数（原问题），
        如果你知道凑出 amount = 10 的最少硬币数（子问题），
        你只需要把子问题的答案加一（再选一枚面值为 1 的硬币）就是原问题的答案。
        因为硬币的数量是没有限制的，所以子问题之间没有相互制，是互相独立的。
        """
        memo = dict()
        # 定义：要凑出金额 n，至少要 dp(n) 个硬币
        def dp(n):
            # init base
            if n == 0:
                return 0
            if n < 0:
                return -1
            inf = float("INF")
            res = inf
            for coin in coins:
                sub = dp(n - coin)
                if sub == -1:  # 子问题无解 跳过
                    continue
                res = min(res, 1 + sub)
            memo[n] = res if res != inf else -1
            return memo[n]

        return dp(amount)

    def coinChange(self, coins, amount: int) -> int:
        # 数组大小为 amount + 1，初始值也为 amount + 1
        # 数组的定义：当目标金额为 i 时，至少需要 dp[i] 枚硬币凑出
        dp = [amount + 1] * (amount + 1)
        # init base case
        dp[0] = 0
        # 定义：要凑出金额 n，至少要 dp(n) 个硬币
        for i in range(0, amount + 1):
            for coin in coins:
                if (i - coin) < 0:
                    # 无解
                    continue
                dp[i] = min(dp[i], 1 + dp[i - coin])
        return dp[amount] if dp[amount] != amount + 1 else -1

    def test(self, coins=[1, 2, 5], amount=11):
        return self.coinChange(coins, amount)


# @lc code=end

